src/plugins/intel_cpu/src/nodes/multinomial.hpp

```diff
@@ -0,0 +1,180 @@
+
+    bool created() const override;
+
+    static bool isSupportedOperation(const std::shared_ptr<const ov::Node>& op, std::string& errorMessage) noexcept;
+
+    bool needPrepareParams() const override;
+    void prepareParams() override;
+
+    bool isExecutable() const override;
+    void execute(dnnl::stream strm) override;
+    void executeDynamicImpl(dnnl::stream strm) override;
+    bool canBeInPlace() const override {
+        return false;
+    }
+
+protected:
+    bool needShapeInfer() const override;
+
+private:
+    /// Multinomial params
+    bool m_with_replacement = false;
+    bool m_log_probs = false;
+    uint64_t m_global_seed = 0;
+    uint64_t m_op_seed = 0;
+
+    /// Shape inference
+    static constexpr size_t PROBS_PORT = 0lu;
+    static constexpr size_t NUM_SAMPLES_PORT = 1lu;
+    static constexpr size_t OUTPUT_PORT = 0lu;
+    bool m_const_inputs[2] = {false, false};
+    VectorDims m_output_shape = {};
+
+    /// General algorithm variables
+    ov::element::Type m_probs_precision;
+    ov::element::Type m_num_samples_precision;
+    ov::element::Type m_output_precision;
+
+    size_t m_probs_count = 0;
+    size_t m_batches_count = 0;
+    size_t m_samples_count = 0;
+    size_t m_samples_probs_count = 0;
+    size_t m_input_elements_count = 0;
+    size_t m_output_elements_count = 0;
+    size_t m_batches_samples_probs_count = 0;
+
+    template <typename P>
+    void execute_types() {
+        switch (m_output_precision) {
+        case ov::element::i32:
+            return execute_internal<P, int32_t>();
+        case ov::element::i64:
+            return execute_internal<P, int64_t>();
+        default:
+            THROW_CPU_NODE_ERR("Multinomial CPU implementation does not support output convert type: ",
+                               m_output_precision);
+        }
+    }
+
+    template <typename P, typename O>
+    void execute_internal() {
+        const auto* probs = reinterpret_cast<const P*>(getParentEdgeAt(PROBS_PORT)->getMemoryPtr()->getData());
+        auto* output = reinterpret_cast<O*>(getChildEdgeAt(OUTPUT_PORT)->getMemoryPtr()->getData());
+
+        std::vector<P> m_cdf(m_input_elements_count);
+        std::vector<P> m_max_per_batch(m_batches_count);
+        std::vector<P> m_random_samples(m_output_elements_count);
+
+        // exp & cumsum
+        if (m_log_probs) {
+            parallel_for(m_batches_count, [&](size_t idx) {
+                auto start_idx = idx * m_probs_count;
+                m_cdf[start_idx] = std::exp(probs[start_idx]);
+                for (size_t i = 1; i < m_probs_count; ++i) {
+                    m_cdf[start_idx + i] = std::exp(probs[start_idx + i]) + m_cdf[start_idx + i - 1];
+                }
+            });
+        } else {
+            parallel_for(m_batches_count, [&](size_t idx) {
+                auto start_idx = idx * m_probs_count;
+                m_cdf[start_idx] = probs[start_idx];
+                for (size_t i = 1; i < m_probs_count; ++i) {
```

Comment:
```suggestion
                for (size_t curr = start_idx + 1, prev = start_idx; curr < (start_idx + m_probs_count); ++curr, ++prev) {
                    m_cdf[curr] = probs[curr] + m_cdf[prev];
                }
```
The loop can be replaced by something similar to this, it will reduce index calculation in loop body by adding offset, index will be just incremented.

Alternative is use std::transform which use two inputs and one output.

Reply:
Replaced with the suggested for loop modification!